common/views/components/Images/Images.js

```diff
@@ -8,87 +8,182 @@ import type {Node} from 'react';
+            })}
+            src={convertImageUri(contentUrl, 30, false)}
+            data-srcset={imageSizes(width).map(size => {
+              return `${convertImageUri(contentUrl, size, false)} ${size}w`;
+            })}
+            sizes={sizesQueries}
+            alt={alt} />
+        }
+
+        {showTasl && <Tasl {...tasl} isFull={isFull} />}
+      </Fragment>
+    );
+  }
+}
 
 export type UiCaptionedImageProps = {|
   ...CaptionedImageProps,
   sizesQueries: string,
   extraClasses?: string,
   preCaptionNode?: Node,
-  maxHeightRestricted?: boolean
+  setTitleStyle?: (value: number) => void
+|}
+
+type UiCaptionedImageState = {|
+  isActive: boolean,
+  computedImageWidth: ?number,
+  isWidthAuto: boolean
 |}
 
-export const CaptionedImage = ({
-  image,
-  caption,
-  sizesQueries,
-  extraClasses = '',
-  preCaptionNode,
-  maxHeightRestricted = false
-}: UiCaptionedImageProps) => {
-  const uiImageProps = {...image, sizesQueries, maxHeightRestricted};
-  return (
-    <figure className={`captioned-image ${extraClasses}`}>
-      <div className='captioned-image__image-container'>
-        {/* https://github.com/facebook/flow/issues/2405 */}
-        {/* $FlowFixMe */}
-        <UiImage {...uiImageProps} />
-      </div>
-      <Caption
-        caption={caption}
-        preCaptionNode={preCaptionNode} />
-    </figure>
-  );
-};
+export class CaptionedImage extends Component<UiCaptionedImageProps, UiCaptionedImageState> {
+  state = {
+    isActive: false,
+    computedImageWidth: null,
+    isWidthAuto: false
+  }
+
+  setComputedImageWidth = (width: number) => {
+    this.props.setTitleStyle && this.props.setTitleStyle(width);
+    this.setState({
+      computedImageWidth: width,
+      isActive: true
+    });
+  }
+
+  setLazyLoaded = () => {
+    this.setState({
+      isWidthAuto: true
+    });
+  }
+
+  render() {
+    const { caption, preCaptionNode, extraClasses, image, sizesQueries } = this.props;
+    const { isActive, computedImageWidth, isWidthAuto } = this.state;
+    const uiImageProps = {...image, sizesQueries};
+
+    return (
+      <figure
+        style={{
+          marginLeft: isActive && isWidthAuto && '50%',
+          transform: isActive && isWidthAuto && computedImageWidth && `translateX(${computedImageWidth / -2}px)`
+        }}
+        className={`captioned-image ${extraClasses || ''}`}>
+        <div
+          style={{
+            display: isWidthAuto && 'inline-block'
+          }}
+          className='captioned-image__image-container relative'>
+          {/* https://github.com/facebook/flow/issues/2405 */}
+          {/* $FlowFixMe */}
+          <UiImage
+            {...uiImageProps}
+            setComputedImageWidth={this.setComputedImageWidth}
+            setLazyLoaded={this.setLazyLoaded} />
+        </div>
+        <Caption
+          width={computedImageWidth}
+          caption={caption}
+          preCaptionNode={preCaptionNode} />
+      </figure>
+    );
+  }
+}
```

Comment:
Could you leave a note as to how this is working, it seems like quite complicated logic, and without tests, it's hard (for me) to fathom.

Reply:
FYI https://github.com/storybooks/storybook/tree/master/addons/jest

Reply:
I've simplified this by having the `CaptionedImage` hold the state that governs e.g. `width: auto` on the image and `display: inline-block` on the image container. Dropped a comment on there, too.